infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -0,0 +1,225 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* A detector that uses ptrace to identify shell injection vulnerabilities. */
+
+/* C standard library */
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <signal.h>
+
+/* POSIX */
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/user.h>
+#include <sys/wait.h>
+
+/* Linux */
+#include <syscall.h>
+#include <sys/ptrace.h>
+
+#include <fstream>
+#include <string>
+#include <vector>
+#include <set>
+
+const std::string kTripWire = "/tmp/tripwire";
+
+#define DEBUG_LOGS 0
+
+#if DEBUG_LOGS
+#define debug_log(...) \
+  do { \
+    fprintf(stderr, __VA_ARGS__); fflush(stdout); \
+    fputc('\n', stderr); \
+  } while (0)
+#else
+#define debug_log(...)
+#endif
+
+#define fatal_log(...) \
+  do { \
+    fprintf(stderr, __VA_ARGS__); \
+    fputc('\n', stderr); \
+    exit(EXIT_FAILURE); \
+  } while (0)
+
+pid_t run_child(char **argv) {
+  // Run the program under test with its args as a child process
+  pid_t pid = fork();
+  switch (pid) {
+    case -1:
+      fatal_log("Fork failed: %s", strerror(errno));
+    case 0:
+      raise(SIGSTOP);
+      execv(argv[1], argv + 1);
+      fatal_log("execv: %s", strerror(errno));
+  }
+  return pid;
+}
+
+std::vector<std::byte> read_memory(pid_t pid, unsigned long long address, size_t size) {
+  std::vector<std::byte> memory;
+
+  for (size_t i = 0; i < size; i += sizeof(long)) {
+    long word = ptrace(PTRACE_PEEKTEXT, pid, address + i, 0);
+    if (word == -1) {
+      return memory;
+    }
+
+    std::byte *word_bytes = reinterpret_cast<std::byte*>(&word);
+    memory.insert(memory.end(), word_bytes, word_bytes+sizeof(long));
+  }
+
+  return memory;
+}
+
+void inspect(pid_t pid, const user_regs_struct &regs) {
+  auto memory = read_memory(pid, regs.rdi, kTripWire.length());
+  if (memory.size() == 0) {
+    return;
+  }
+
+  std::string path(reinterpret_cast<char*>(
+        memory.data()), std::min(memory.size(), kTripWire.length()));
+  debug_log("inspecting");
+  if (path == kTripWire) {
+    kill(pid, SIGKILL);
+    printf("===BUG DETECTED: Shell injection===\n");
+    // TODO: Get/print stacktrace.
+    exit(1);
+  }
+}
+
+void trace(std::set<pid_t> pids) {
+  while (!pids.empty()) {
+    std::vector<pid_t> new_pids;
+
+    auto it = pids.begin();
+    while (it != pids.end()) {
+      auto pid = *it;
+      int status = 0;
+
+      int result = waitpid(pid, &status, __WALL | WNOHANG);
+      if (result == -1) {
+        it = pids.erase(it);
+        continue;
+      }
+
+      if (result == 0) {
+        // Nothing to report yet.
+        ++it;
+        continue;
+      }
+
+      debug_log("finished waiting %d", pid);
+
+      if (WIFEXITED(status) || WIFSIGNALED(status)) {
+        debug_log("%d exited", pid); 
+        it = pids.erase(it);
+        continue;
+      }
+
+      bool is_syscall = WIFSTOPPED(status) && WSTOPSIG(status) == (SIGTRAP | 0x80);
```

Comment:
What is 0x80? Maybe put this in a const.

Reply:
added a comment to explain this.